xdsl/dialects/pdl.py

```diff
@@ -0,0 +1,323 @@
+
+        if not isinstance(self.attributes['name'], StringAttr):
+            raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, args: list[SSAValue]) -> ApplyNativeConstraintOp:
+        return ApplyNativeConstraintOp.build(
+            result_types=[],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class ApplyNativeRewriteOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlapply_native_rewrite-mlirpdlapplynativerewriteop
+    """
+    name: str = "pdl.apply_native_rewrite"
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OpAttr[StringAttr]
+    args: Annotated[VarOperand, AnyPDLType]
+    res: Annotated[VarOpResult, AnyPDLType]
+
+    def verify_(self) -> None:
+        if 'name' not in self.attributes:
+            raise Exception("ApplyNativeRewriteOp requires a 'name' attribute")
+
+        if not isinstance(self.attributes['name'], StringAttr):
+            raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, args: list[SSAValue],
+            result_types: list[Attribute]) -> ApplyNativeRewriteOp:
+
+        return ApplyNativeRewriteOp.build(
+            result_types=[result_types],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class AttributeOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlattribute-mlirpdlattributeop
+    """
+    name: str = "pdl.attribute"
+    value: OptOpAttr[Attribute]
+    value_type: Annotated[OptOperand, TypeType]
+    output: Annotated[OpResult, AttributeType]
+
+
+@irdl_op_definition
+class EraseOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlerase-mlirpdleraseop
+    """
+    name: str = "pdl.erase"
+    op_value: Annotated[Operand, OperationType]
+
+
+@irdl_op_definition
+class OperandOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperand-mlirpdloperandop
+    """
+    name: str = "pdl.operand"
+    value_type: Annotated[OptOperand, TypeType]
+    value: Annotated[OpResult, ValueType]
+
+
+@irdl_op_definition
+class OperandsOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperands-mlirpdloperandsop
+    """
+    name: str = "pdl.operands"
+    value_type: Annotated[Operand, RangeType[TypeType]]
+    value: Annotated[OpResult, RangeType[ValueType]]
+
+
+@irdl_op_definition
+class OperationOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperation-mlirpdloperationop
+    """
+    name: str = "pdl.operation"
+    opName: OptOpAttr[StringAttr]
+    attributeValueNames: OpAttr[ArrayAttr[StringAttr]]
+
+    operandValues: Annotated[VarOperand, ValueType | RangeType[ValueType]]
+    # in PDL docs, this is just a handle to AttributeType, not a range.
```

Comment:
What do you mean by this?

Reply:
This comment was from @martin-luecke, I'm not sure exactly what he meant. The PDL documentation says that it's a handle, but I'm not sure what type that's supposed to be. Would be good to have some insight on that also.

Reply:
I think this is related to the autogenerated documentation.
By looking at the TableGen file, this should not be a range, just a Variadic of AttributeType!

Reply:
Ah ok, makes sense

Reply:
In the docs it says for `attributeValues`: 
Description: "PDL handle to an mlir::Attribute", which looks incorrect to me, as there can be multiple attributes, not just a single one.
So I assume this should be just Variadic of AttributeType. 
The question that opens up here is: Why is this designed differently to `operandValues`, where I can supply a pdl.range of operands. So why can I not supply a pdl.range of attributes? This should work with attributeValueNames as long as the cardinality of both is equal.

So this comment was me not understanding why they modeled it differently without an apparent good reason + their incorrect documentation. I think we should completely remove this comment and also remove `RangeType[AttributeType]]` from `attributeValues` to match their design.